4.4 - Add fatal error handling to ExceptionTrap

markstory opened the pull request:
Rough in fatal error handling to ExceptionTrap. This may need further revisions once I do more extensive integration testing in an application.

Refs #16228

markstory committed:
Add fatal error handling to ExceptionTrap

Rough in fatal error handling to ExceptionTrap. This may need further
revisions once I do more extensive integration testing in an
application.

Co-authored-by: othercorey <corey.taylor.fl@gmail.com>

othercorey commented on src/Error/ExceptionTrap.php:
```diff
@@ -185,6 +185,83 @@ public function handleException(Throwable $exception): void
+     */
+    public function handleShutdown(): void
+    {
+        $megabytes = $this->_config['extraFatalErrorMemory'] ?? 4;
```
Why isn't this in KB like the helper function?

markstory replied:
If we change the unit size we'll make upgrading harder as the config keys are shared between the two systems. This has to move anyways as I'm pretty convinced this needs to happen during registration, and at this point it is too late as we could be handling an OOM error and no longer have any space to use for the page rendering process.

markstory replied:
I was thinking about this more, and our strategy is better if you're able to change memory limits at runtime, and your physical host isn't over-subscribed in memory with no swap.

With our implementation no memory is used up on every single request just in case a fatal error happens. Instead, when an error happens we borrow a bit more RAM to handle the error.

othercorey commented on src/Error/ExceptionTrap.php:
```diff
@@ -185,6 +185,83 @@ public function handleException(Throwable $exception): void
+        }
+        $limit = trim($limit);
+        $units = strtoupper(substr($limit, -1));
+        $current = (int)substr($limit, 0, strlen($limit) - 1);
```
```suggestion
        $current = (int)substr($limit, 0, -1);
```

othercorey commented on src/Error/ExceptionTrap.php:
```diff
@@ -185,6 +185,83 @@ public function handleException(Throwable $exception): void
+        $limit = trim($limit);
+        $units = strtoupper(substr($limit, -1));
+        $current = (int)substr($limit, 0, strlen($limit) - 1);
+        if ($units === 'M') {
```
To be 100% compatible, this should use all digits as the limit and then compare the remaining string with defaulting to bytes.

https://www.php.net/manual/en/faq.using.php#faq.using.shorthandbytes

markstory committed:
Add coverage for increaseMemoryLimit

markstory committed:
Fix typing and phpcs errors.

markstory committed:
4.4 - ExceptionTrap::instance() (#16278)

Add global state to ExceptionTrap

I don't really like global state, but in this situation we have natural
global state that interfaces with PHP. Having an `instance()` method
will allow middleware to get the current globally registered instance.

This may seem a bit dumb as we could make another ExceptionTrap in the
middleware. My concern with this that developers would need to add their
callbacks twice if they were using that feature which is also new.

There is an argument to remove callback support as userland code could
build a renderer/logger that handle their needs.

Co-authored-by: othercorey <corey.taylor.fl@gmail.com>
